### PR TITLE
[fix](memory) Fix BE OOM when load -238 fail

### LINKED ISSUE #12661: [Bug] BE OOM when load fails
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Version

ead016e0d2567030c297277c1dca63198cc18efb

### What's Wrong?
When the load fails, the memory will increase instantly

After the reason load fails, flush will also fail, `wait flush` will return immediately, and then you can continue to add batches to memtable.

image

### What You Expected?

no oom

### How to Reproduce?

_No response_

### Anything Else?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by xinyiZzz**
# Proposed changes

Issue Number: close #12661

## Problem summary

When the flush is triggered when the load channel exceeds the mem limit, if the flush fails, an error message is returned and the load is terminated.

Usually flush failure is -238 error code. Because the memtable is frequently flushed after the load channel exceeds the mem limit, the number of segments exceeds the max value.

Before fix:
image

After fix:
image

## Checklist(Required)

1. Does it affect the original behavior: 
    - [x] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [x] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/runtime/tablets_channel.cpp [hunk 1]
```diff
@@ -237,8 +238,22 @@ Status TabletsChannel::reduce_mem_usage(int64_t mem_limit) {
...
+                    "tablet writer failed to reduce mem consumption by flushing memtable, "
+                    "tablet_id=$0, txn_id=$1, err=$2, errcode=$3, msg:$4",
+                    writers[i]->tablet_id(), _txn_id, st.code(), st.precise_code(),
+                    st.get_error_msg());
+            LOG(WARNING) << err_msg;
+            PTabletError* error = tablet_errors->Add();
+            error->set_tablet_id(writers[i]->tablet_id());
+            error->set_msg(err_msg);
+            _broken_tablets.insert(writers[i]->tablet_id());
+        }
```
Comment (zhannngchen) [inline be/src/runtime/tablets_channel.cpp:256]: If some error happened while `flush_memtable_and_wait`, we should not call `wait_flush` in L260, in that situation, `wait_flush` might wait infinitely?

Comment (xinyiZzz) [reply be/src/runtime/tablets_channel.cpp:256]: `wait_flush` should not wait infinitely, and the flush token that fails to submit the flush will not actually wait.

Do not return error directly, pass the error to the sink to decide whether to terminate the load.

**Review by zhannngchen — APPROVED** [review]
LGTM

**Review by dataroaring — APPROVED** [review]
LGTM

File: be/src/runtime/tablets_channel.cpp (hunk 1 above)
Comment (xinyiZzz) [reply be/src/runtime/tablets_channel.cpp:256]: done, `_broken_tablets` does not `wait flush`

**Pull request merged**